### ISSUE TITLE
How to Save Pruned Models with Modified Val.py

### ISSUE THREAD

**Issue by EveMyadzePike** [ultralytics/yolov5#6728]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I followed the pruning/sparsity tutorial, pruned a model to 30% sparsity and the output was:
![prune](https://user-images.githubusercontent.com/76539970/154950955-ddf31c26-3553-42a1-a92a-f9a9253c377e.jpg)

I tried saving the model using the code:

  torch.save({'model':deepcopy(de_parallel(model)).half()}, pruned)

But I get an error message:
![error](https://user-images.githubusercontent.com/76539970/154951055-7fbdf20a-6733-403a-9a47-cfbf494fdc12.jpg)

How can I correctly modify val.py to save the pruned model?

**Comment by glenn-jocher** [comment]
@EveMyadzePike example model saving code is in train.py:
https://github.com/ultralytics/yolov5/blob/dbbb57cf0b8004dddb7e3bc4506b68b44b60896a/train.py#L388-L407

**Comment by EveMyadzePike** [comment]
Thank you for responding! I used the example code and modified train.py, but then to test that I had correctly saved the pruned model, I ran val.py (unmodified) on my saved pruned model. I got different results, so I thought my save process was incorrect.

**Comment by EveMyadzePike** [comment]
With 30% pruning, I got these results:
![prune](https://user-images.githubusercontent.com/76539970/154963748-e280c684-d340-44d9-9b22-5e78cc49b127.jpg)

I saved my pruned model and ran an unmodified version of val.py. The results are different.
![results](https://user-images.githubusercontent.com/76539970/154964123-d0fca7ed-37c6-485a-b5c0-3f158ddbf684.jpg)

**Comment by PraveenMNaik** [comment]
> Thank you for responding! I used the example code and modified train.py, but then to test that I had correctly saved the pruned model, I ran val.py (unmodified) on my saved pruned model. I got different results, so I thought my save process was incorrect.

Please let me know... how to save pruned model... pls share here the lines of code......

**Comment by glenn-jocher** [comment]
@PraveenMNaik see the checkpoint saving code in train.py:
https://github.com/ultralytics/yolov5/blob/7215a0fb41a90d8a0bf259fa708dff608a1f0262/train.py#L386-L399

**Comment by PraveenMNaik** [comment]
@glenn-jocher... true its is implemented in tain.py. But when we prune using val.py the above code segment will not be used.... how to integrate it in val.py...

**Comment by bryanbocao** [comment]
Hello @PraveenMNaik, just wanted to check if you solve this issue? I am also facing a similar one. Thanks!

> > Thank you for responding! I used the example code and modified train.py, but then to test that I had correctly saved the pruned model, I ran val.py (unmodified) on my saved pruned model. I got different results, so I thought my save process was incorrect.
> 
> Please let me know... how to save pruned model... pls share here the lines of code......

@glenn-jocher @EveMyadzePike

**Comment by glenn-jocher** [comment]
Hello @bryanbocao, thank you for reaching out about this issue. As `val.py` is not responsible for model checkpoint saving during pruning, it is not necessary to modify this script. If you have already pruned your model and saved the checkpoint using the provided code in `train.py`, then you can directly use the pruned model for inference or other operations. Note that when you use a pruned model, the resulting detections might be different from the ones obtained with the original model. During the pruning process, some weights were set to zero, which affects the network's ability to make accurate predictions. Please let us know if you need further assistance with this issue.

**Comment by bryanbocao** [comment]
@glenn-jocher Thanks for checking.

```If you have already pruned your model and saved the checkpoint using the provided code in `train.py`,``` as mentioned by me and other developers,  @PraveenMNaik, simply following the format to save the pruned model still caused issues
```
 # Save model 
 if (not nosave) or (final_epoch and not evolve):  # if save 
     ckpt = { 
         'epoch': epoch, 
         'best_fitness': best_fitness, 
         'model': deepcopy(de_parallel(model)).half(), 
         'ema': deepcopy(ema.ema).half(), 
         'updates': ema.updates, 
         'optimizer': optimizer.state_dict(), 
         'wandb_id': loggers.wandb.wandb_run.id if loggers.wandb else None, 
         'date': datetime.now().isoformat()} 
```

**Comment by glenn-jocher** [comment]
Hello @bryanbocao, I understand that you encountered issues when trying to save a pruned model using the provided code in `train.py`. Thank you for letting us know. Could you please provide more information on the issues you encountered, such as any error messages or abnormal behaviors you noticed? This would help us in identifying the root cause of the problem and find solutions to it. In the meantime, you can also refer to the pruning/sparsity tutorial in the YOLOv5 Wiki and double-check that all the required steps have been followed correctly. Thank you!

**Comment by bryanbocao** [comment]
Hi @glenn-jocher, thanks for taking your time with this issue. I moved forward so I will try to rephrase what we did today. If I have time later I will log the error messages for reproducing the issue.

In ```val.py```, we did
```
prune(model, 0.5)
torch.save(model, 'pruned.pt')
```

By inspecting the weights of model after prune(), we verified that nearly haft of them are zeros. But then there is an error saying that something related to IO Text cannot be serialized, which is wrapped in the DetectMultiBackends module.

Then we did something like
```
prune(model, 0.5)
ckpt = { 
         'epoch': epoch, 
         'best_fitness': None, 
         'model': deepcopy(de_parallel(model)).half(), 
         'ema': deepcopy(ema.ema).half(), 
         'updates': None, 
         'optimizer': None, 
         'wandb_id': None, 
         'date': datetime.now().isoformat()} 
torch.save(ckpt, 'pruned.pt')
```
but still got errors.

Later we solved it by
```
prune(model, 0.5)
ckpt = { 
         'epoch': epoch, 
         'best_fitness': None, 
         'model': model.model, # Notice here
         'ema': deepcopy(ema.ema).half(), 
         'updates': None, 
         'optimizer': None, 
         'wandb_id': None, 
         'date': datetime.now().isoformat()} 
torch.save(ckpt, 'pruned.pt')
```

We tested we can load the saved 'pruned.pt' and run ```val.py```.

Basically, we carefully check the format of model (DetectMultiBackend or yolo.model) saved and wrapped in ```DetectMultiBackend``` and the format where we actually save the model.

**Comment by glenn-jocher** [comment]
Hello @bryanbocao, thank you for giving me an update. I appreciate the effort you put to troubleshoot the issue and for sharing your experience with us. Your proposed solution to the issue seems reasonable and it is great to hear that it is working for you. As you noticed, the unexpected behavior was caused by the serialization of the IO Text. In general, it is better to serialize only what is necessary, such as the model architecture, its parameters, and the optimizer's state, and not to include unwanted objects like debug information or partial models. When you save the checkpoint, double-check that the format of the model and the parameters are correct and that they match the expected input format in your downstream script. If you encounter any further issues, feel free to let us know. Thanks again for using YOLOv5 and happy coding!

**Comment by PraveenMNaik** [comment]
@bryanbocao could you be able to get reduction in model size after prune operation? or same model size?

**Comment by bryanbocao** [comment]
@PraveenMNaik Haven't tried it in this case. But in general, it uses unstructural pruning which is **not** supposed to reduce model size - just masking unimportant weights by setting them to zeros.

**Comment by abhinavvsharma** [comment]
Hi @glenn-jocher, I am a collaborator with @bryanbocao. Here is how we reached to our solution:
Upon checking the datatype of the saved model file ```yolov5n.pt``` using ```type(mode.pt)``` we get that the file is a dictionary of format:
```
{ 
         'epoch': epoch, 
         'best_fitness': None, 
         'model': <class 'models.yolo.Model'>,  # Notice here
         'ema': None, 
         'updates': None, 
         'optimizer': None, 
         'wandb_id': None, 
         'date': datetime.now().isoformat()
} 
```
In an attempt to further understand the attributes of ```model['model']``` using ```vars(model['model']``` we get the result as:
```
{
'training': False, 
'_parameters': OrderedDict(), 
'_buffers': OrderedDict(), 
'_non_persistent_buffers_set': set(), 
'_backward_hooks': OrderedDict(), 
'_is_full_backward_hook': None, 
'_forward_hooks': OrderedDict(), 
'_forward_pre_hooks': OrderedDict(), 
'_state_dict_hooks': OrderedDict(), 
'_load_state_dict_pre_hooks': OrderedDict(),
'_modules': OrderedDict([('model', Sequential(#Model Architecture)]), 
'yaml_file': 'yolov5n.yaml', 
'yaml': {'nc': 80, 'depth_multiple': 0.33, 'width_multiple': 0.25, 'anchors': [[10, 13, 16, 30, 33, 23], [30, 61, 62, 45, 59, 119], [116, 90, 156, 198, 373, 326]], 'backbone': [[-1, 1, 'Conv', [64, 6, 2, 2]], [-1, 1, 'Conv', [128, 3, 2]], [-1, 3, 'C3', [128]], [-1, 1, 'Conv', [256, 3, 2]], [-1, 6, 'C3', [256]], [-1, 1, 'Conv', [512, 3, 2]], [-1, 9, 'C3', [512]], [-1, 1, 'Conv', [1024, 3, 2]], [-1, 3, 'C3', [1024]], [-1, 1, 'SPPF', [1024, 5]]], 'head': [[-1, 1, 'Conv', [512, 1, 1]], [-1, 1, 'nn.Upsample', ['None', 2, 'nearest']], [[-1, 6], 1, 'Concat', [1]], [-1, 3, 'C3', [512, False]], [-1, 1, 'Conv', [256, 1, 1]], [-1, 1, 'nn.Upsample', ['None', 2, 'nearest']], [[-1, 4], 1, 'Concat', [1]], [-1, 3, 'C3', [256, False]], [-1, 1, 'Conv', [256, 3, 2]], [[-1, 14], 1, 'Concat', [1]], [-1, 3, 'C3', [512, False]], [-1, 1, 'Conv', [512, 3, 2]], [[-1, 10], 1, 'Concat', [1]], [-1, 3, 'C3', [1024, False]], [[17, 20, 23], 1, 'Detect', ['nc', 'anchors']]], 'ch': 3}, 
'save': [4, 6, 10, 14, 17, 20, 23],
'names': ['person'], 
'inplace': True, 'stride': tensor([ 8., 16., 32.]), 
'nc': 80, 
'hyp': {'lr0': 0.01, 'lrf': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'warmup_epochs': 3.0, 'warmup_momentum': 0.8, 'warmup_bias_lr': 0.1, 'box': 0.05, 'cls': 0.006250000000000001, 'cls_pw': 1.0, 'obj': 1.0, 'obj_pw': 1.0, 'iou_t': 0.2, 'anchor_t': 4.0, 'fl_gamma': 0.0, 'hsv_h': 0.015, 'hsv_s': 0.7, 'hsv_v': 0.4, 'degrees': 0.0, 'translate': 0.1, 'scale': 0.5, 'shear': 0.0, 'perspective': 0.0, 'flipud': 0.0, 'fliplr': 0.5, 'mosaic': 1.0, 'mixup': 0.0, 'copy_paste': 0.0, 'label_smoothing': 0.0}, 'class_weights': tensor([1.])
}
```

On checking the datatype of model after,
```
prune(model, 0.5)
torch.save(model, 'pruned.pt')
```
we see that it is a ```<class 'DetectMultiBackend'>```. This is the reason of the error [here](https://github.com/ultralytics/yolov5/issues/6728#:~:text=Then%20we%20did%20something%20like). ```torch.save()``` expects a variable with ```<class 'models.yolo.Model'>``` datatype here. ```<class 'DetectMultiBackend'>``` has attribute ```.model``` i.e., ```model.model``` that actually contains the data expected by dictionary input of ```torch.save()```. Therefore, the final solution posted by @bryanbocao results in saving the pruned model file successfully.

**Comment by glenn-jocher** [comment]
Thank you for the detailed explanation, @abhinavvsharma. Your investigation and solution provide a clear understanding of the issue and the steps taken to resolve it. It's great to see how diligently you've investigated the root cause, enabling the successful saving of the pruned model. This information will undoubtedly be helpful for others encountering similar issues. If you have any more discoveries or require further assistance, feel free to reach out. Happy coding!
